### PR TITLE
HBASE-22637 fix flaky test in TestMetaTableMetrics

### PR DESCRIPTION
**Pull request by symat**
the problem was seen only on branch 2.0 and 2.1, but it should be unrelated to the minor version of the branches and the same fix can be applied on all 2.x branches

fixes / changes:
- wait until all the metrics appear on JMX (previously we were only waiting for the bean to appear, I think this was the cause behind fleakyness)
- testing all the jmx attributes for all metrics / operations
- rename the test (test -> testMetaTableMetricsInJmx)

### CODE REVIEW

**Comment by symat** [comment]
if the test is fixed on the master branch, then I also plan to create an other pull request for branch 1 and 1.4 (as this patch an not be built on branch 1, only after some minor changes)

File: hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestMetaTableMetrics.java [hunk 1]
```diff
@@ -109,46 +112,109 @@ public static void setupBeforeClass() throws Exception {
...
+        containsPositiveJmxAttributesFor("MetaTable_region_.+_lossy_request"),
+        containsPositiveJmxAttributesFor("MetaTable_table_" + NAME1 + "_request"),
+        containsPositiveJmxAttributesFor("MetaTable_client_.+_put_request"),
+        containsPositiveJmxAttributesFor("MetaTable_client_.+_get_request"),
+        containsPositiveJmxAttributesFor("MetaTable_client_.+_delete_request"),
+        containsPositiveJmxAttributesFor("MetaTable_client_.+_lossy_request")
+      ).matches(jmxMetrics);
+
+      if (allMetricsFound) {
+        LOG.info("all the meta table metrics found with positive values: " + jmxMetrics.toString());
```
Comment (petersomogyi) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestMetaTableMetrics.java]: On branch-2 we have slf4j so we could benefit from using parameterized logging.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestMetaTableMetrics.java [hunk 2]
```diff
@@ -162,18 +228,22 @@ private void writeData(Table t) throws IOException {
...
+      Map<String, Double> existingAttrs = new HashMap<>();
       for (MBeanAttributeInfo attrInfo : beanInfo.getAttributes()) {
-        existingAttrs.add(attrInfo.getName());
+        Object value = mb.getAttribute(target, attrInfo.getName());
+        if (attrInfo.getName().startsWith(METRICS_ATTRIBUTE_NAME_PREFIX)
+            && value instanceof Number) {
+          existingAttrs.put(attrInfo.getName(), Double.parseDouble(value.toString()));
+        }
       }
+      LOG.warn("MBean Found: " + target);
```
Comment (petersomogyi) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/coprocessor/TestMetaTableMetrics.java]: Should this be info level?

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 183 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 301 | master passed |
| +1 | compile | 68 | master passed |
| +1 | checkstyle | 85 | master passed |
| +1 | shadedjars | 341 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 278 | master passed |
| +1 | javadoc | 41 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 294 | the patch passed |
| +1 | compile | 68 | the patch passed |
| +1 | javac | 68 | the patch passed |
| +1 | checkstyle | 86 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 341 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 925 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 301 | the patch passed |
| +1 | javadoc | 42 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 16043 | hbase-server in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 19819 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/346 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 437700cf29e1 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / b20044c612 |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/1/testReport/ |
| Max. process+thread count | 4765 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/1/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 191 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | hbaseanti | 0 |  Patch does not have any anti-patterns. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ master Compile Tests _ |
| +1 | mvninstall | 333 | master passed |
| +1 | compile | 71 | master passed |
| +1 | checkstyle | 89 | master passed |
| +1 | shadedjars | 362 | branch has no errors when building our shaded downstream artifacts. |
| +1 | findbugs | 275 | master passed |
| +1 | javadoc | 44 | master passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 317 | the patch passed |
| +1 | compile | 70 | the patch passed |
| +1 | javac | 70 | the patch passed |
| +1 | checkstyle | 86 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedjars | 349 | patch has no errors when building our shaded downstream artifacts. |
| +1 | hadoopcheck | 989 | Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2. |
| +1 | findbugs | 295 | the patch passed |
| +1 | javadoc | 41 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 16725 | hbase-server in the patch failed. |
| +1 | asflicense | 26 | The patch does not generate ASF License warnings. |
| | | 20657 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.regionserver.TestRegionReplicasWithModifyTable |
|   | hadoop.hbase.client.TestAsyncTableAdminApi |
|   | hadoop.hbase.util.TestFromClientSide3WoUnsafe |
|   | hadoop.hbase.client.TestAsyncTableAdminApi3 |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/346 |
| Optional Tests |  dupname  asflicense  javac  javadoc  unit  findbugs  shadedjars  hadoopcheck  hbaseanti  checkstyle  compile  |
| uname | Linux 7c99152b21a2 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | master / 0c8dc5d97e |
| maven | version: Apache Maven 3.5.4 (1edded0938998edf8bf061f1ceb3cfdeccf443fe; 2018-06-17T18:33:14Z) |
| Default Java | 1.8.0_181 |
| findbugs | v3.1.11 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/2/testReport/ |
| Max. process+thread count | 4638 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-346/2/console |
| Powered by | Apache Yetus 0.9.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by jatsakthi** [comment]
The changed looks great! Tested it locally, ran the test 10 times without failure. +1 from me as well. Thanks @symat

**Pull request merged**